### PR TITLE
[c10d] Fix color value for comm split being negative

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137544
* #138374
* __->__ #137855

Fixes https://github.com/pytorch/pytorch/issues/137856.

### Issue 1
Today under `ProcessGroupNCCL::Options`, color is declared as:
```
    int64_t split_color{0};
```
When passing this variable to `ncclCommSplit` which accepts `int`, the value may overflow and become negative, as in #137856. But NCCL API only accepts non-negative colors (or `NCCL_SPLIT_NOCOLOR`).

But that's not all.

### Issue 2
`split_color` is pybind'ed to python frontend. If we just change from `int64_t` to `int` in C++, pybind will complain:
```
[rank0]: TypeError: (): incompatible function arguments. The following argument types are supported:
[rank0]:     1. (self: torch._C._distributed_c10d.ProcessGroupNCCL.Options, arg0: int) -> None
```
This is because python `int` represents a wider range than C++ `int`. So we cannot pass hash values -- which are potentially big ints -- from python to C++. The PR modulo the hash value with `c_int`'s max value.

cc @XilunWu @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

File: torch/csrc/distributed/c10d/NCCLUtils.cpp [hunk 1]
```diff
@@ -82,12 +90,22 @@ std::shared_ptr<NCCLComm> NCCLComm::split(
       ncclCommSplit(sourceComm, color_id, rank, &(comm->ncclComm_), &config),
       sourceComm, // wait on parent comm
       std::nullopt);
+  if (color_id >= 0) {
+    // Waiting for parent comm above still does not seem to guarantee the child
+    // comm ptr is valid. Therefore we add a manual wait here for safety.
+    // TODO: remove this wait after NCCL fix the semantics.
+    while (!comm->ncclComm_) {
+      C10D_SCHED_SLEEP();
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/NCCLUtils.cpp:98]: hm, possible infinite sleep?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp [hunk 2]
```diff
@@ -486,7 +486,18 @@ class TORCH_API ProcessGroupNCCL : public Backend {
     // Optional "parent" backend and color to create communicators from
     // via `ncclCommSplit`
     std::shared_ptr<ProcessGroupNCCL> split_from;
-    int64_t split_color{0};
+    // Color to use for `ncclCommSplit`, values:
+    // * Non-negative value: in group;
+    // * NCCL_SPLIT_NOCOLOR (-1): not in group;
+    // * NCCL_SPLIT_NOCOLOR - 1: uninitialized.
+    // [Note 1]: the type must be `int` instead of `int64_t` because NCCL API
+    // accepts int. Otherwise, an imlicit conversion may happen at the API call
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:494]: typo implicit

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -4475,26 +4476,38 @@ def _create_process_group_wrapper(
...
     global _world
     if use_hashed_name:
-        pg_name = _hash_ranks(ranks)
-        while pg_name in _world.pg_names.values():
-            pg_name = hashlib.sha1(bytes(pg_name + "_", "utf-8")).hexdigest()
+        pg_name = _hash_ranks_to_str(ranks)
     else:
         pg_name = str(_world.group_count)
         _world.group_count += 1
+    # TODO: why is group count incremented only in the else path?
```
Comment (wconstab) [inline torch/distributed/distributed_c10d.py:4510]: also where is _world.pg_names updated?  It doesn't look like it gets updated when we create a new name here, but you seem to rely on its length as a proxy for 'group_count'

File: torch/csrc/distributed/c10d/NCCLUtils.cpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/NCCLUtils.cpp:98]: Good call. Can I defer a jail break to the PR on top where we have a definite timeout value?
Today the default value is -1, which means no timeout.
I agree the deferral is not nice, but there is quite some work to pull part of the PR on top down here.

File: torch/distributed/distributed_c10d.py (hunk 3 above)
Comment (kwen2501) [reply torch/distributed/distributed_c10d.py:4510]: It is updated here:
https://github.com/pytorch/pytorch/blob/47e40455669251282286ceadad2d30643f45963c/torch/distributed/distributed_c10d.py#L1978

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp (hunk 2 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:494]: Will fix in later PR.

File: torch/csrc/distributed/c10d/NCCLUtils.cpp (hunk 1 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/NCCLUtils.cpp:98]: Addressed in https://github.com/pytorch/pytorch/pull/138374

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp (hunk 2 above)
Comment (kwen2501) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:494]: Addressed in https://github.com/pytorch/pytorch/pull/138374

**Review by wconstab — APPROVED** [review]
thanks for adding the timeout. I think the timeout is too long but i am not sure what value is ok.

**Comment by wdvr** [comment]
revert -m 'reverting since this is failing internal tests, maybe a NCCL version mismatch between oss and fbcode?' -c regression

**Comment by wdvr** [comment]
ignore above, forward fix added in https://github.com/pytorch/pytorch/pull/138488

**Pull request closed without merging**